### PR TITLE
Rem 219

### PR DESCRIPTION
**Pull request by FatimaSalami**
fiscal year
is a multiTab form..
you can check the behaviour from asp and compare it please
second tab display depends on period and other selections of first form

 a change in table component is done by the help of ali mokdad (the status issue we discussed in meeting already).. if another solution is better we can have discussion..
 please any reviewer pass on some other pages that include status and test the table or external grid please

### CODE REVIEW

**Comment by Ghinwa-Arab** [comment]
-fiscal year should light red on validation
-add months label
image

**Comment by Ghinwa-Arab** [comment]
minimize the width of drop down
image

File: src/pages/fiscal-years/index.js [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    {
+      field: 'endDate',
+      headerName: _labels.endDate,
+      flex: 1,
+      type: 'date'
+    }
+  ]
+
+  const add = () => {
+    openForm('')
```
Comment (Ghinwa-Arab) [inline src/pages/fiscal-years/index.js:67]: no need to pass ''

File: src/pages/fiscal-years/index.js [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+  }
+
+  function openForm(recordId) {
+    stack({
+      Component: FiscalYearWindow,
+      props: {
+        labels: _labels,
+        maxAccess: access,
+        recordId: recordId ? recordId : null
```
Comment (Ghinwa-Arab) [inline src/pages/fiscal-years/index.js:76]: no need for `recordid?recordId:null`, enough to pass recordId

File: src/pages/fiscal-years/index.js [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+      title: _labels.fiscalYear
+    })
+  }
+
+  const edit = obj => {
+    openForm(obj.fiscalYear)
+  }
+
+  const del = async obj => {
+    await postRequest({
```
Comment (Ghinwa-Arab) [inline src/pages/fiscal-years/index.js:88]: add try catch

File: src/pages/fiscal-years/Windows/FiscalYearWindow.js [hunk 4]
```diff
@@ -0,0 +1,34 @@
+// ** Custom Imports
```
Comment (Ghinwa-Arab) [inline src/pages/fiscal-years/Windows/FiscalYearWindow.js:1]: remove comment

File: src/pages/fiscal-years/Windows/FiscalYearWindow.js [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+  const tabs = [{ label: labels.fiscalYear }, { label: labels.periods, disabled: !store.recordId }]
+
+  return (
+    <>
+      <CustomTabs tabs={tabs} activeTab={activeTab} setActiveTab={setActiveTab} />
+      <CustomTabPanel index={0} height={height} value={activeTab}>
+        <FiscalYearForm labels={labels} maxAccess={maxAccess} editMode={editMode} setStore={setStore} store={store} />
+      </CustomTabPanel>
+      <CustomTabPanel height={height} index={1} value={activeTab}>
```
Comment (Ghinwa-Arab) [inline src/pages/fiscal-years/Windows/FiscalYearWindow.js:27]: why u're passing height? it's necessary?

File: src/pages/fiscal-years/Windows/FiscalYearWindow.js [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    recordId: recordId || null
+  })
+
+  const tabs = [{ label: labels.fiscalYear }, { label: labels.periods, disabled: !store.recordId }]
+
+  return (
+    <>
+      <CustomTabs tabs={tabs} activeTab={activeTab} setActiveTab={setActiveTab} />
+      <CustomTabPanel index={0} height={height} value={activeTab}>
+        <FiscalYearForm labels={labels} maxAccess={maxAccess} editMode={editMode} setStore={setStore} store={store} />
```
Comment (Ghinwa-Arab) [inline src/pages/fiscal-years/Windows/FiscalYearWindow.js:25]: remove `editMode={editMode} `not used

File: src/pages/fiscal-years/forms/FiscalYearForm.js [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+  const { platformLabels } = useContext(ControlContext)
+  const { recordId } = store
+
+  const invalidate = useInvalidate({
+    endpointId: SystemRepository.FiscalYears.page
+  })
+
+  const { formik } = useForm({
+    initialValues: {
+      recordId: null,
```
Comment (Ghinwa-Arab) [inline src/pages/fiscal-years/forms/FiscalYearForm.js:32]: do we have recordId here?, in table u're passing rowId=fiscalYear

File: src/pages/fiscal-years/forms/FiscalYearForm.js [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+    enableReinitialize: false,
+    validateOnChange: true,
+    validationSchema: yup.object({
+      fiscalYear: yup
+        .number()
+        .transform((value, originalValue) => validateNumberField(value, originalValue))
+        .min(1900, labels.yearMin)
+        .max(2099, labels.yearMax),
+      startDate: yup.string().required(' '),
+      endDate: yup.string().required(' '),
```
Comment (Ghinwa-Arab) [inline src/pages/fiscal-years/forms/FiscalYearForm.js:49]: remove empty strings

File: src/pages/fiscal-years/forms/FiscalYearForm.js [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+    onSubmit: async obj => {
+      const fiscalYear = obj.fiscalYear
+      await postRequest({
+        extension: SystemRepository.FiscalYears.set,
+        record: JSON.stringify(obj)
+      })
+
+      if (!recordId) {
+        toast.success(platformLabels.Added)
+        formik.setValues({
```
Comment (Ghinwa-Arab) [inline src/pages/fiscal-years/forms/FiscalYearForm.js:62]: replace by formik.setFieldValue

File: src/pages/fiscal-years/forms/FiscalYearForm.js [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+          ...obj,
+          recordId: fiscalYear
```
Comment (Ghinwa-Arab) [inline src/pages/fiscal-years/forms/FiscalYearForm.js:64]: set fiscalYear returned from api response not from object

File: src/pages/fiscal-years/forms/FiscalYearForm.js [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+          <Grid container spacing={4}>
+            <Grid item xs={12}>
+              <CustomTextField
+                name='fiscalYear'
+                label={labels.fiscalYear}
+                value={formik.values.fiscalYear}
+                onChange={formik.handleChange}
+                onClear={() => formik.setFieldValue('fiscalYear', '')}
+                error={formik.touched.fiscalYear && Boolean(formik.errors.fiscalYear)}
+                helperText={formik.touched.fiscalYear && formik.errors.fiscalYear}
```
Comment (Ghinwa-Arab) [inline src/pages/fiscal-years/forms/FiscalYearForm.js:112]: remove helper text

File: src/pages/fiscal-years/forms/FiscalYearForm.js [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+            <Grid item xs={12}>
+              <ResourceComboBox
+                datasetId={DataSets.FY_PERIOD_STATUS}
+                name='status'
+                label={labels.status}
+                valueField='key'
+                displayField='value'
+                values={formik.values}
+                onChange={(event, newValue) => {
+                  if (newValue) {
```
Comment (Ghinwa-Arab) [inline src/pages/fiscal-years/forms/FiscalYearForm.js:171]: use ternary operator (shorthand)

File: src/pages/fiscal-years/forms/FiscalYearForm.js [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+            <Grid item xs={12}>
+              <ResourceComboBox
+                datasetId={DataSets.FY_PERIODS}
+                name='periods'
+                label={labels.periods}
+                valueField='key'
+                displayField='value'
+                values={formik.values}
+                onChange={(event, newValue) => {
+                  if (newValue) {
```
Comment (Ghinwa-Arab) [inline src/pages/fiscal-years/forms/FiscalYearForm.js:150]: use ternary operator (shorthand)

File: src/pages/fiscal-years/forms/PeriodsForm.js [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+      periods: obj.map(({ id, periodId, ...rest }) => ({
+        periodId: id,
+        ...rest
+      }))
+    }
+    postRequest({
+      extension: SystemRepository.FiscalPeriodPack.set2,
+      record: JSON.stringify(data)
+    })
+      .then(res => {
```
Comment (Ghinwa-Arab) [inline src/pages/fiscal-years/forms/PeriodsForm.js:39]: can we use async await

File: src/pages/fiscal-years/forms/PeriodsForm.js [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+    },
+    maxAccess,
+    enableReinitialize: false,
+    validateOnChange: true,
+    validationSchema: yup.object({
+      periods: yup
+        .array()
+        .of(
+          yup.object().shape({
+            statusName: yup.string().required(' '),
```
Comment (Ghinwa-Arab) [inline src/pages/fiscal-years/forms/PeriodsForm.js:59]: remove empty strings

File: src/pages/fiscal-years/forms/PeriodsForm.js [hunk 16]
```diff
@@ -0,0 +1,185 @@
...
+          width: 600,
+
+          title: labels?.period
+        })
+      }
+    }
+  ]
+
+  useEffect(() => {
+    recordId && getPeriods()
```
Comment (Ghinwa-Arab) [inline src/pages/fiscal-years/forms/PeriodsForm.js:129]: use `if(recordId)` instead `recordId &&`

File: src/pages/fiscal-years/forms/PeriodsForm.js [hunk 17]
```diff
@@ -0,0 +1,185 @@
...
+  }, [recordId])
+
+  const getPeriods = () => {
+    const defaultParams = `_fiscalYear=${recordId}`
```
Comment (Ghinwa-Arab) [inline src/pages/fiscal-years/forms/PeriodsForm.js:133]: `const defaultParams = `_fiscalYear=${recordId}`
 var parameters = defaultParams`
 remove and assign `_fiscalYear=${recordId} `directly to parameters

File: src/pages/fiscal-years/forms/PeriodsForm.js [hunk 18]
```diff
@@ -0,0 +1,185 @@
...
+    var parameters = defaultParams
+    getRequest({
+      extension: SystemRepository.Period.qry,
+      parameters: parameters
+    })
+      .then(res => {
```
Comment (Ghinwa-Arab) [inline src/pages/fiscal-years/forms/PeriodsForm.js:139]: can we use async await instead

File: src/pages/fiscal-years/forms/PeriodsForm.js [hunk 19]
```diff
@@ -0,0 +1,185 @@
...
+  ]
+
+  useEffect(() => {
+    recordId && getPeriods()
+  }, [recordId])
+
+  const getPeriods = () => {
+    const defaultParams = `_fiscalYear=${recordId}`
+    var parameters = defaultParams
+    getRequest({
```
Comment (Ghinwa-Arab) [inline src/pages/fiscal-years/forms/PeriodsForm.js:135]: add try catch

File: src/pages/fiscal-years/forms/PeriodsModuleForm.js [hunk 20]
```diff
@@ -0,0 +1,172 @@
...
+import { CommonContext } from 'src/providers/CommonContext'
+import { ControlContext } from 'src/providers/ControlContext'
+
+const PeriodsForm = ({ recordId, labels, maxAccess, row, window, editMode }) => {
+  const { getRequest, postRequest } = useContext(RequestsContext)
+  const { getAllKvsByDataset } = useContext(CommonContext)
+  const { platformLabels } = useContext(ControlContext)
+
+  const post = obj => {
+    const data = {
```
Comment (Ghinwa-Arab) [inline src/pages/fiscal-years/forms/PeriodsModuleForm.js:21]: use try catch and can we use async await?

File: src/pages/fiscal-years/forms/PeriodsModuleForm.js [hunk 21]
```diff
@@ -0,0 +1,172 @@
...
+          { from: 'value', to: 'statusName' }
+        ],
+        columnsInDropDown: [{ key: 'value', value: 'Name' }],
+        displayFieldWidth: 2
+      }
+    }
+  ]
+
+  useEffect(() => {
+    recordId && getModules()
```
Comment (Ghinwa-Arab) [inline src/pages/fiscal-years/forms/PeriodsModuleForm.js:85]: use `if(recordId)`

File: src/pages/fiscal-years/forms/PeriodsModuleForm.js [hunk 22]
```diff
@@ -0,0 +1,172 @@
...
+  }, [recordId])
+
+  const getModules = () => {
+    ;(async function () {
+      const moduleData = await getAllModules()
+      const defaultParams = `_fiscalYear=${recordId}&_periodId=${row.periodId}`
```
Comment (Ghinwa-Arab) [inline src/pages/fiscal-years/forms/PeriodsModuleForm.js:91]: remove 
`const defaultParams = `_fiscalYear=${recordId}&_periodId=${row.periodId}`
var parameters = defaultParams`
      
 and pass `_fiscalYear=${recordId}&_periodId=${row.periodId}` directly

File: src/pages/fiscal-years/forms/PeriodsModuleForm.js [hunk 23]
```diff
@@ -0,0 +1,172 @@
...
+      var parameters = defaultParams
+      getRequest({
+        extension: SystemRepository.FiscalModule.qry,
+        parameters: parameters
+      })
+        .then(res => {
```
Comment (Ghinwa-Arab) [inline src/pages/fiscal-years/forms/PeriodsModuleForm.js:97]: use async await if possible

File: src/pages/fiscal-years/forms/PeriodsModuleForm.js [hunk 24]
```diff
@@ -0,0 +1,172 @@
...
+          })
+
+          const mappedModules = modules.map(({ id, ...rest }, index) => {
+            return {
+              id: index + 1,
+              ...rest
+            }
+          })
+
+          formik.setValues({
```
Comment (Ghinwa-Arab) [inline src/pages/fiscal-years/forms/PeriodsModuleForm.js:127]: use formik.setFieldValue('modules',mappedModules)

File: src/pages/fiscal-years/forms/FiscalYearForm.js (hunk 7 above)
Comment (FatimaSalami) [reply src/pages/fiscal-years/forms/FiscalYearForm.js:32]: needed to be passed to infobutton 
i am placing fiscal year value in it

File: src/pages/fiscal-years/forms/FiscalYearForm.js (hunk 10 above)
Comment (FatimaSalami) [reply src/pages/fiscal-years/forms/FiscalYearForm.js:64]: it is not returned on post

**Comment by imtithalhalalo** [comment]
in edit mode when I clicked Periods it shows like that 
Screenshot 2024-08-07 144046

File: src/pages/fiscal-years/forms/PeriodsForm.js [hunk 25]
```diff
@@ -0,0 +1,182 @@
...
+          ...rest
+        }))
+      }
+      await postRequest({
+        extension: SystemRepository.FiscalPeriodPack.set2,
+        record: JSON.stringify(data)
+      })
+
+      toast.success(platformLabels.Edited)
+      getPeriods()
```
Comment (imtithalhalalo) [inline src/pages/fiscal-years/forms/PeriodsForm.js:42]: needs await before it

File: src/pages/fiscal-years/forms/PeriodsForm.js [hunk 26]
```diff
@@ -0,0 +1,182 @@
...
+      }
+    },
+    {
+      component: 'button',
+      name: 'saved',
+      onClick: (e, row) => {
+        stack({
+          Component: PeriodsModuleForm,
+          props: {
+            labels: labels,
```
Comment (imtithalhalalo) [inline src/pages/fiscal-years/forms/PeriodsForm.js:114]: you can pass them as labels, maxAccess, editMode

**Pull request merged**